python/ql/lib/semmle/python/dataflow/new/internal/DataFlowPrivate.qll

```diff
@@ -168,6 +168,47 @@ private predicate synthDictSplatArgumentNodeStoreStep(
   )
 }
 
+/**
+ * Holds if `nodeFrom` is the value yielded by the `yield` found at `nodeTo`.
+ *
+ * For example, in
+ * ```python
+ * for x in l:
+ *   yield x.name
+ * ```
+ * data from `x.name` is stored into the `yield` (and can subsequently be read out of the iterable).
+ */
+predicate yieldStoreStep(Node nodeFrom, Content c, Node nodeTo) {
+  exists(Yield yield, Function func |
+    nodeTo.asCfgNode() = yield.getAFlowNode() and
+    nodeFrom.asCfgNode() = yield.getValue().getAFlowNode() and
+    func.containsInScope(yield)
+  |
+    exists(Comp comp | func = comp.getFunction() |
+      (
+        comp instanceof ListComp or
+        comp instanceof GeneratorExp
+      ) and
+      c instanceof ListElementContent
+      or
+      comp instanceof SetComp and
+      c instanceof SetElementContent
+      or
+      comp instanceof DictComp and
+      c instanceof DictionaryElementAnyContent
+    )
+    or
+    not exists(Comp comp | func = comp.getFunction()) and
+    (
+      c instanceof ListElementContent
+      or
+      c instanceof SetElementContent
+      or
+      c instanceof DictionaryElementAnyContent
+    )
```

Comment:
I wonder if we really need `SetElementContent` or `DictionaryElementAnyContent` here? 

My mental model is that when you call a generator function you get an iterable back, so from my understanding you would always need to turn those elements into a set/dict before use anyway.

That is, you could do `set(my_generator_func())` but doing `my_set | my_generator_func()` results in an error. (and likewise for dictionaries).

Reply:
Our model of the `dict` constructor does not convert content, so that would have to change in order to model something like
```
dict([k, v for v in l])
```

I did check if we were always just adding list content and the constructor pulled it out and I would also be open to that regime.

Reply:
I tried that out, it seems a lot nicer and also improved our coverage slightly..

Reply:
I'm surprised that https://github.com/github/codeql/pull/17577/commits/38b1eb7c71a0dd77b59e493d3222a56f13c5ca31 is enough. Do we already have the logic in place for sets?

I'm also surprised that dict-comprehensions and set-comprehensions just works :thinking: -- is that because a they are transformed into `dict(_comp_function())` and `set(_comp_function())`?

Reply:
Yes the `set` constructor already handles list elements: https://github.com/github/codeql/blob/main/python/ql/lib/semmle/python/frameworks/Stdlib.qll#L4315-L4317

Reply:
I think that raw dict comprehensions do not actually work: https://github.com/github/codeql/blob/main/python/ql/test/library-tests/dataflow/coverage/test.py#L182-L184

Reply:
could we add a comment saying that this might need to be revised when adding support for dict-comprehensions then? :pray:

Reply:
we discussed together that it might make sense to do some followup work on how dict-comprehensions are handled by the extractor, and what it would take to support them better in our analysis.